source/server/listener_impl.cc

```diff
@@ -220,6 +223,38 @@ Server::DrainManager& ListenerFactoryContextBaseImpl::drainManager() { return *d
 // Must be overridden
 Init::Manager& ListenerFactoryContextBaseImpl::initManager() { NOT_IMPLEMENTED_GCOVR_EXCL_LINE; }
 
+FilterChainRebuilder::FilterChainRebuilder(
+    ListenerImpl& listener, const envoy::config::listener::v3::FilterChain* const& filter_chain)
+    : listener_(listener), filter_chain_(filter_chain), rebuild_complete_(false),
+      rebuild_init_manager_(std::make_unique<Init::ManagerImpl>("rebuild_init_manager")),
+      rebuild_watcher_("rebuild_watcher", [this] {
+        rebuild_complete_ = true;
+        callbackToWorkers();
+      }) {}
+FilterChainRebuilder::~FilterChainRebuilder() = default;
+
+void FilterChainRebuilder::addRebuildInfo(FilterChainRebuildInfoPtr rebuild_info) {
+  std::string worker_name = rebuild_info->worker_name_;
+  std::string listener_name = rebuild_info->listener_name_;
+
+  // Add this listener_name to the callback set of this worker.
+  listeners_on_worker_to_callback_[worker_name].insert(listener_name);
+}
+
+void FilterChainRebuilder::callbackToWorkers() {
+  bool success = true;
+  if (!success) {
+    return;
+  }
+  // Find all matching workers and listeners, send callback.
+  for (auto& worker_listeners : listeners_on_worker_to_callback_) {
+    const auto& worker_name = worker_listeners.first;
+    auto worker_ptr = listener_.getWorkerByName(worker_name);
+    worker_ptr->notifyListeners(filter_chain_);
+  }
```

Comment:
On completion of filter chain rebuilding, callback to workers to retry connections using this filter chain.